FileCache: Clean up empty directories on remove() #3162

afbora opened the pull request:
## Describe the PR

When the cache is removed, it moves to the root directory, removing empty directories.

## Related issues
<!-- PR relates to issues in the `kirby` repo or ideas on `feedback.getkirby.com`: -->

- Fixes #3162 

## Ready?
<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Added in-code documentation (if needed)
- [x] CI passes (runs automatically when the PR is created or run `composer ci` locally)
  Running locally requires PHPUnit, PHP-CS-Fixer, Psalm, PHPCPD and PHPMD.

<!-- We will take care of the following TODO when reviewing the PR. -->

- [x] Checked whether the PR needs documentation, if needed added to the [release docs checklist](https://github.com/getkirby/getkirby.com/pulls)

afbora committed:
FileCache: Clean up empty directories on remove() #3162

lukasbestle commented on src/Cache/FileCache.php:
```diff
@@ -178,10 +179,34 @@ public function remove(string $key): bool
+    }
+
+    /**
+     * Removes empty directories safely by moving to the root directory.
```
```suggestion
     * Removes empty directories safely by checking each directory
     * up to the root directory
```

lukasbestle commented on src/Cache/FileCache.php:
```diff
@@ -178,10 +179,34 @@ public function remove(string $key): bool
+
+                if (empty($files) === true) {
+                    if (Dir::remove($dir) === true) {
+                        $this->removeEmptyDirectories(dirname($dir));
```
The recursive approach will work quite well for shallow nesting levels (structures just a few levels deep). But if cache IDs contain potentially dozens of slashes (for whatever reason), this implementation will require too much memory for its purpose IMO.

We can fix this by using an iterative approach instead. Untested example code you could try:

```php
// ensure the path doesn't end with a slash for the next comparison
$dir = rtrim($dir, '/\/');

while (Str::startsWith($dir, $this->root()) === true && $dir !== $this->root()) {
    $files = array_diff(scandir($dir) ?? [], ['.', '..']);

    if (empty($files) === true && Dir::remove($dir) === true) {
        // continue with the next level up
        $dir = dirname($dir);
    }

    // no need to continue with the next level up as $dir was not deleted
    return;
}
```

afbora replied:
It was a 99% working code. Copy & paste & test 😅 Thank you 👍

afbora committed:
Use iterative approach on removing empty directories #3162

Since there may be a performance problem, an iterative solution that suggested by @lukasbestle  has been applied instead of recursive.

afbora committed:
Add "not empty dirs" unit test for FileCache::removeEmptyDirectories #3162

afbora committed:
Add missing FileCache::removeEmptyDirectories() description in doc block #3162

lukasbestle reviewed (approved):
Awesome! 👍
